File: multinationalCurrency/src/money/Sum.java
package money;

public class Sum implements Expression {
    Expression augend;
    Expression addend;
    Sum(Expression augend, Expression addend){
        this.addend = addend;
        this.augend = augend;
    }
    public Expression times(int multoplier) {
        return new Sum(augend.times(multoplier), addend.times(multoplier));
    }
    public Expression plus(Expression addend) {
        return new Sum(this, addend);
    }

    public Money reduce(Bank bank, String to){
        int amount = augend.reduce(bank, to).amount + addend.reduce(bank, to).amount;
        return new Money(amount, to);
    }
}

File: multination_currency_rb/sum.rb
# frozen_string_literal: true

require_relative 'money'

class Sum
  attr_accessor :augend, :added

  def initialize(augend, added)
    @augend = augend
    @added = added
  end

  def reduce(bank, to)
    amount = @augend.reduce(bank, to).amount + @added.reduce(bank, to).amount
    Money.new(amount, to)
  end

  def plus(added)
    Sum.new(self, added)
  end
end
